### DISCUSSION TITLE
I don't understand the Quick tour of HASPmota

### DISCUSSION THREAD

**Discussion by richardsg307** [arendst/Tasmota#20228] [Q&A] [answered]
I'm trying to get a Lilygo T-Display-S3 to display temperatures and other things related to my boiler and got drawn down the Berry LGVL route. I've had some success but am stuck on clearing the old values from the screen before writing the new (every 5 seconds).
Now I've been drawn into the HASPmota thing because looking at the Quick Tour showed screens similar to what I'm aiming for.
But I'm totally confused and that might be resolved by a few questions being answered.
1) I loaded the haspmota_demo.tapp file and it ran overlaying on the screen on top of what I had in my autoexec.be file. I then cleared the autoexec file and now just get the demo on screen. Is that expected behaviour?
2) No values are being displayed (eg Temperature guage is blank). I take it i don't have the right sensors configured but do have 2 x DS18x20 coming in on GPIO43. Can I amend the demo file to read one of them to display on the temperature guage? 
3) When I look at the demo file in Notepad++ it's huge and looks like a mixture of Berry (?) code and machine language that comes up as random symbols. Buried in the middle of it, I can see the JSON block that looks like the 'Understanding the Template' section. The file just doesn't stack up with the 'Running HASPmota' instructions. What's going on?
4) Is there a tool available for creating the JSON Templates?
5) Is there somewhere that give a clear guide this? Or a YouTube video?
6) Knowing my use case described above can anyone give me some pointers as to the way to go?

Edit: I've taken the plunge and deleted the demo file, created the simple autoexec.be as described and made pages.jsonl with just the status bar, the Temp gauge and the XYZs. That seems to work although still without any values.
7) What do I need to change to get it to report on the screen the 2 temperatures coming in from my DS18x20s? They show up in the Console logs as DS18B20-1 and -2. I tried replacing the "text_rule":"MPU9250#AZ" with "text_rule":"DS18B20-1" but that doesn't work.
PS. I don't have the touch screen version so am limited to just page 1.

**Comment by sfromis** [comment]
If you use `haspmota_demo.tapp`, this will be started with Tasmota, and run at the same time as anything else you may be doing. You could also run the same demo without the `.tapp` by having the demo `pages.jsonl` in the file system (see later), and in `autoexec.be`:
```
import haspmota
haspmota.start(false, "/pages.jsonl")
```

The demo temperature gauge will only be active if you enable the mostly useless ESP32 internal "temperature" sensor using `SetOption146`. This also serves an an example of how you can process sensors on the same device as the display, look for `ESP32#Temperature` in the demo `pages.jsonl`.

If you want to look at the demo, you can unpack it as it is just an uncompressed zip archive, renamed to `.tapp`, but you do not need to do this as most of it is the same as the example detailed and explained in the docs.
https://tasmota.github.io/docs/HASPmota/#understanding-the-template
You can also find the full content of the demo here:
https://github.com/arendst/Tasmota/tree/development/tasmota/berry/haspmota_src/haspmota_demo
And the main pages specifically:
https://github.com/arendst/Tasmota/blob/development/tasmota/berry/haspmota_src/haspmota_demo/pages.jsonl

Tasmota as such has no tools for creating the template JSON (which is quite feasible to do in a simple text editor), but any JSON based editor tool may be helpful. Maybe there also exists tools to help with creating the same data for [openHASP](https://www.openhasp.com/) which the `.jsonl` format "is heavily inspired from". Searching for openHASP conent should also be useful for HASPmota, as long as you only look for what's in the `.jsonl` files. In addition to this, Tasmota also has a number of features where you can use the Berry language to  update page content and so on.

If your use case is using locally connected sensors, the demo technique for `ESP32#Temperature` shows how it can be done, without any Berry coding. If the use case is showing data from another Tasmota device, you can subscribe to get values via MQTT, unpick the value from the payload, and then code like this:
```
var temperature = 22.3
global.p1b2.text = f'Temperature: {temperature:%.1f}'
```
to update page 1, id 2 with the formatted text, including a temperature with one decimal place.
You can also update various attributes the same way (basically same attributes you already give in the `.jsonl` file) using the same way with `global.p1b2.text_color` and so on.

Do note that outside of the "very similar" `.jsonl` format, HASPmota and openHASP are entirely separate projects, working in different ways, and possibly not 100% compatible in all cases. Still, jsonl-constructs from openHASP should generally work in HASPmota also, differences may be bugs which can be fixed in HASPmota.

**Reply by richardsg307** [reply]
Thank you. I have done the SetOption thing and got the ESP32 Temperature displaying now. I've also followed the instructions and have it running with just the simple autoexec.be file and a pages.jsonl with just the page 1 stuff on it (and the demo file deleted), So that makes more sense.
I will have a look at the OpenHASP stuff noting what you say about limiting it to the JSON bits.
I still haven't resolved displaying my DS18x20 temperatures yet.
What you say about picking up data from MQTT is useful as I have Home Assistant running with an ESP8266 currently doing the sensing at my boiler. HA calculates other things like 'run time today' and 'oil cost today', so it would be good if I could report that back to the Lilygo ESP32 Display.

**Reply by sfromis** [reply]
Picking up data from a locally connected DS19B20 would be using the trigger `DS18B20#Temperature` (adjust sensor name if you have used aliases, or have more than one sensor), instead of where you see `ESP32#Temperature` in the demo code. You need to match exactly what you see in the sensor JSON data already being published by Tasmota.

If you have data in HA you want to push to the display, a rudimentary method would be to push assignments like `global.p1b2.text = value` by publishing such a snippet of Berry code as argument to the `Br` command, executed via MQTT. Other methods include registering a Tasmota command name in Berry, if you want to avoid having the display-specific stuff in HA, or whatever you could envision by doing vt coding in Berry.

**Reply by richardsg307** [reply]
That works, thank you so much. I have 2 x DS18x20 in my breadboard (3 in the boiler) so used "DS18D20-1#Temperature" and "DS18D20-2#Temperature".
It's a bit weird but now I have the lines
```
{"id":10,"obj":"label","x":172,"y":35,"w":140,"text":"Boiler Status","align":0}
{"id":11,"obj":"label","x":172,"y":55,"w":140,"text":"Outflow = ","align":0,"text_rule":"DS18B20-1#Temperature","text_rule_format":"x=%5.2f","text_rule_formula":"val"}
{"id":12,"obj":"label","x":172,"y":75,"w":140,"text":"Inflow = ","align":0,"text_rule":"DS18B20-2#Temperature","text_rule_format":"y=%5.2f","text_rule_formula":"val"}
{"id":13,"obj":"label","x":172,"y":75,"w":140,"text":"Cabinet = ","align":0,"text_rule":"DS18B20-2#Temperature","text_rule_format":"y=%5.2f","text_rule_formula":"val"}
```
in my pages.jsonl file, when the screen starts up the Inflow/Outflow text appears on the screen, the Cabinet line is missing and after a few seconds when the values appear, Inflow and Outflow change to Y= and X= respectively. Like I need to clear a cache or something!
Anyway, I'll work on it now I've made progress with your help.

**Reply by sfromis** [reply, marked as answer]
It probably helps if you do not have two objects at the same x/y position. And better include your legend as part of the formatting, or create a separate static label field for only the legend.

**Reply by richardsg307** [reply]
Well all I can say is Doh and Doh! That's the trouble when you just copy and paste blindly. All sorted now and moving on finalising the screen layout and getting the MQTT feed of data working with Home Assistant. Big thanks.
